fragment.go

```diff
@@ -1454,88 +1465,79 @@ func (f *fragment) bulkImport(rowIDs, columnIDs []uint64, options *ImportOptions
-	// first we'll try the "small update" path. If there aren't many bits being
-	// set, it isn't worth it to do a full import and snapshot. Instead we
-	// leverage individual set/clear bits which get set in memory and appended
-	// to the op log.
+	var localBitmap *roaring.Bitmap
+	smallWrite := false
+	f.mu.Lock()
+	defer f.mu.Unlock() // TODO the big import path doesn't need to acquire the lock this soon.
 	if len(columnIDs)+f.opN < f.MaxOpN {
-		for i := range rowIDs {
-			rowID, columnID := rowIDs[i], columnIDs[i]
-			if options.Clear {
-				_, err = f.clearBit(rowID, columnID)
-			} else {
-				_, err = f.setBit(rowID, columnID)
-			}
-			if err != nil {
-				return errors.Wrapf(err, "importing %dth bit clear:%v", i, options.Clear)
-			}
-		}
-		// forcibly recalculate the cache - setbit just calls invalidate, but in
-		// order to maintain parity with the "normal" bulkimport path, we want
-		// to recalculate it at the end of the import
-		f.cache.Recalculate()
-		return nil
-	} // end "small update" path - after this is the real bulk import path
-
-	// Create a temporary bitmap which will be populated by rowIDs and columnIDs
-	// and then merged into the existing fragment's bitmap.
-	localBitmap := roaring.NewBitmap()
+		smallWrite = true
+		// If the number of operations is small, we'll write directly to local storage
+		localBitmap = f.storage
+	} else {
+		// Create a temporary bitmap which will be populated by rowIDs and columnIDs
+		// and then merged into the existing fragment's bitmap.
+		localBitmap = roaring.NewBitmap()
 
-	// Disconnect op writer so we don't append updates.
-	localBitmap.OpWriter = nil
+		// Disconnect op writer so we don't append updates.
+		localBitmap.OpWriter = nil
+	}
 
-	// rowSet maintains the set of rowIDs present in this import.
-	// It allows the cache to be updated once per row, instead of once
-	// per bit.
+	// rowSet maintains the set of rowIDs present in this import. It allows the
+	// cache to be updated once per row, instead of once per bit. TODO: consider
+	// sorting by rowID/columnID first and avoiding the map allocation here. (we
+	// could reuse rowIDs to store the list of unique row IDs)
 	rowSet := make(map[uint64]struct{})
 	lastRowID := uint64(0)
 
-	// Process every bit by writing to a local bitmap,
-	// to be merged with fragment storage next.
-	for i := range rowIDs {
+	// replace columnIDs with calculated positions to avoid allocation.
+	for i := 0; i < len(columnIDs); i++ {
 		rowID, columnID := rowIDs[i], columnIDs[i]
-
-		// Determine the position of the bit in the storage.
 		pos, err := f.pos(rowID, columnID)
 		if err != nil {
 			return err
 		}
-
-		// Write to local storage.
-		_, err = localBitmap.Add(pos)
-		if err != nil {
-			return err
-		}
-
-		// Reduce the StatsD rate for high volume stats
-		f.stats.Count("ImportBit", 1, 0.0001)
+		columnIDs[i] = pos
 
 		// Add row to rowSet.
 		if i == 0 || rowID != lastRowID {
```

Comment:
you could also use a lastRowID which can't validly occur -- I think doing lastRowID-- after creating it will make it a value that i don't think can actually meaningfully work in our system, and that simplifies the test here.